src/index.ts

```diff
@@ -115,44 +147,46 @@ export default function useCreateReducerWithEffect(
   return [state, dispatch];
 }
 
-export function composeReducers(reducers) {
-  return (state, action) => {
+export function composeReducers<S, A>(
+  reducers: ReducerWithSideEffects<S, A>[]
+) {
+  return (state: S, action: A) => {
     let reducerCount = reducers.length;
-    let sideEffects = [];
+    let sideEffects: SideEffect<S, A>[] = [];
     let noUpdateCount = 0;
 
-    const reducedResult = reducers.reduceRight((prevState, reducer, index) => {
-      // This is to handle the asymmetry in the useReducerWithSideFffect API.
-      // Whereas regular reducers have a consistent API that is state in, state out
-      // useReducerWithSideEffects has state in, state+sideEffects out
-      let state;
-      if (index === reducerCount - 1) {
-        state = prevState;
-      } else {
-        state = prevState.state;
-      }
-
-      const result = reducer(state, action);
+    const reducedResult = reducers.reduceRight(
+      (prevState, reducer) => {
+        // This is to handle the asymmetry in the useReducerWithSideFffect API.
+        // Whereas regular reducers have a consistent API that is state in, state out
+        // useReducerWithSideEffects has state in, state+sideEffects out
+        const result = reducer(prevState.state, action);
 
-      let returnValue;
-      // When we do not have an update, then we increment our no-update counter and
-      // return the previous state.
-      if (result === NO_UPDATE_SYMBOL) {
-        noUpdateCount++;
+        let returnValue;
+        // When we do not have an update, then we increment our no-update counter and
+        // return the previous state.
+        if (result === NO_UPDATE_SYMBOL) {
+          noUpdateCount++;
 
-        returnValue = {
-          state,
-        };
-      } else {
-        returnValue = result;
-      }
+          returnValue = {
+            state,
+          };
+        } else {
+          returnValue = result;
+        }
 
-      if (result && Array.isArray(result.sideEffects)) {
-        sideEffects = sideEffects.concat(result.sideEffects);
-      }
+        if (
+          result &&
+          result !== NO_UPDATE_SYMBOL &&
+          Array.isArray(result.sideEffects)
+        ) {
+          sideEffects = sideEffects.concat(result.sideEffects);
+        }
 
-      return returnValue;
-    }, state);
+        return returnValue;
+      },
+      { state, sideEffects: [] }
```

Comment:
Instead of having to deal with different prevState types (state, state + sideEffects), I made them all the same here with a default empty sideEffects array.

Reply:
this creates a new object which I believe may cause other issues